infra/Pos/Recovery/Info.hs

```diff
@@ -2,24 +2,88 @@
 -- and provide helpers related to it.
 
 module Pos.Recovery.Info
-       ( MonadRecoveryInfo(..)
+       ( SyncStatus (..)
+       , MonadRecoveryInfo(..)
+       , recoveryInProgress
        , recoveryCommGuard
        ) where
 
-import           System.Wlog (WithLogger, logDebug)
 import           Universum
 
+import qualified Data.Text.Buildable
+import           Formatting          (bprint, build, sformat, stext, (%))
+import           System.Wlog         (WithLogger, logDebug)
+
+import           Pos.Core            (HasCoreConstants, SlotCount, SlotId, slotIdF,
+                                      slotSecurityParam)
+
+-- | An algebraic data type which represents how well we are
+-- synchronized with the network.
+data SyncStatus
+    = SSDoingRecovery
+    -- ^ We are doing recovery right now, so it's better to wait until
+    -- we finish.
+    | SSUnknownSlot
+    -- ^ We don't know current slot, so we are definitely not
+    -- synchronized well enough.
+    | SSLagBehind { sslbTipSlot     :: !SlotId
+                  -- ^ Our tip's slot (if our tip is genesis, we use 0
+                  -- as local slot).
+                  , sslbCurrentSlot :: !SlotId }
+    -- ^ We know current slot, but our tip's slot lags behind current
+    -- slot too much.
+    | SSKindaSynced
+    -- ^ We are kinda synchronized, i. e. all previously described
+    -- statuses are not about us.
+
+instance Buildable SyncStatus where
+    build =
+        \case
+            SSDoingRecovery -> "we're doing recovery"
+            SSUnknownSlot -> "we don't know current slot"
+            SSLagBehind {..} ->
+                bprint
+                    ("we lag behind too much, our tip's slot is: " %slotIdF %
+                     ", but current slot is " %slotIdF)
+                    sslbTipSlot
+                    sslbCurrentSlot
+            SSKindaSynced -> "we are moderately synchronized"
+
 class Monad m => MonadRecoveryInfo m where
-    -- | Returns if 'RecoveryHeader' is 'Just' (which is equivalent to “we're
-    -- doing recovery”).
-    recoveryInProgress :: m Bool
+    -- | Returns our sycnrhonization status. The argument determines
+    -- how much we should lag behind for 'SSLagBehind' status to take
+    -- place. See 'SyncStatus' for details.
+    -- Implementation must check conditions in the same order as they
+    -- are enumerated in 'SyncStatus'.
+    getSyncStatus :: SlotCount -> m SyncStatus
+
+-- | Returns if 'RecoveryHeader' is 'Just' (which is equivalent to “we're
```

Comment:
are you sure you didn't forget to update comment?

Reply:
It's still correct.

Reply:
maybe this statement is true implicitly, but IMHO it's better to clarify it in the comment

Reply:
Previously it was true implicitly too (i. e. there was no guarantee that it will be implemented using `RecoveryHeader`).
But I can change it, ok.